proposals/default_optional_arguments/proposal.txt

```diff
@@ -141,8 +142,36 @@ A use case of this behavior of the function present() could be:
       ...
     end function foo
 
+5.2. intent(out)
+
+There are several different possible behaviors for intent(out)
+variables. One option is to disallow default values on optional
+intent(out) variables. However, there are use cases where it is
+desirable to return the value of a variable if a caller requests it,
+without requiring the added logic of present() and "temporary"
+variables described above.
+
+Alternatively, we might treat intent(in) and intent(out) variables
+identically, so that the following would be valid:
+
+    real function foo(a)
+      real, intent(out), optional :: a = 0
+    end function foo
+
+The benefit is that, just like proposed for intent(in) variables,
+intent(out) variables would be usable regardless of whether the value
+is actually passed back to the caller, removing the need for present()
+and "temporary" variables. The complication is that a present
+intent(out) variable is initially undefined, meaning the logic of
+present() is still necessary in this case. Furthermore, the behavior
+suggested above where present() = .true. always for variables with a
+specified default would not be desired for intent(out) variables.
+
+A third option is that optional intent(out) variables with a specified
+default are always initialized to that default, even when present,
+rather than leaving them undefined.
```

Comment:
I am in favor of such a behavior. However, what would the behavior be for:

```fortran
subroutine foo(a)
  real, intent(out), allocatable, optional :: a(:, :) = 0
end subroutine foo
```

Reply:
This is a very good question, and I think applies to `intent(in)` and `intent(inout)` arguments too (if no actual argument is given). I’ll add a section 5.3 on allocatable arrays, it definitely should be addressed. Off the top of my head, one option is of course to forbid defaults for allocatables. This isn’t preferable, but I suspect allowing defaults for scalars only would still address the majority of use cases. Another is to forbid array broadcasting as a default specifier. This would mean `a(:) = [0,0]` would be permissible, but not `a(:) = 0`. Any other ideas?

Reply:
It is indeed also true for `intent(in)` and `intent(inout)`. I have no knowlegde about compiler development, so I have no idea what would be possible to implement or not. The easiest thing is probably to forbid default values for allocatable arrays, but it could complicate the standard.